README.md
# Pod Billing Service

توابع این ماژول قابلیت های زیر را در اختیار شما قرار می دهد:

* تمامی چرخه مربوط به فاکتور از ایجاد، پرداخت، تایید تا ابطال آن
* روش های پرداخت (لینک های پرداخت کیف پول و درگاه، پرداخت نامتعارف)
* ایجاد و مدیریت فاکتورهای تسهیمی
* تسویه حساب کسب و کار
* ایجاد و مدیریت بن های تخفیف
* ایجاد و مدیریت برداشت مستقیم از حساب بانکی

CHANGELOG.md
# Pod Billing Service

## 2.0.0 ##
* APIs are called through service call
* _token_ & _token_issuer_ fields changed to token & tokenIssuer
* dealer related functions moved to pod-dealing-service module

## 1.3.0 ##
* Voucher APIs added
* Direct debate APIs added

## 1.2.0 ##
* Token issuer can be set as TOKEN_ISSUER in environment variables
* Updating pod-base-service

## 1.1.1 ##
* Updating pod-base-service

## 1.1.0 ##
* SERVER_TYPE system variable has to be set
* General APIs added: getOtt, getGuildList, getCurrencyList

## 1.0.1 ##
* در هنگام وارد کردن قیمت می تواند از عبارت auto نیز استفاده کند


lib/types.js
// Missing Objects: 25, 26, 27, 35, 36, 37, 38

/**
 * @typedef {object} confObj
 * @property {string} apiToken
 * @property {string} tokenIssuer
 */

/**
 * @typedef {object} productIdListObj
 * @property {integer} productId
 * @property {integer} price
 * @property {integer} quantity
 * @property {string} productDescription
 */

/**
 * @typedef {object} invoiceItemListObj
 * @property {integer} invoiceItemId
 * @property {integer} price
 * @property {integer} quantity
 * @property {string} productDescription
 */

// #1
/**
 * @typedef {object} issueInvoiceObj
 * @property {string} [redirectURL]
 * @property {integer} [userId]
 * @property {string} [billNumber]
 * @property {string} [description]
 * @property {string} [deadline]
 * @property {productIdListObj[]} productIdList
 * @property {string} guildCode
 * @property {string} [currencyCode]
 * @property {integer} [addressId]
 * @property {string} [voucherHash]
 * @property {string} [preferredTaxRate]
 * @property {boolean} [verificationNeeded]
 * @property {boolean} [verifyAfterTimeout]
 * @property {boolean} [preview]
 * @property {string} [metadata]
 * @property {string} [safe]
 * @property {string} [postVoucherEnabled]
 * @property {string} [hasEvent]
 * @property {string} [eventTitle]
 * @property {string} [eventTimeZone]
 * @property {object[]} [eventReminders]
 * @property {string} [eventDescription]
 * @property {object} [eventMetadata]
 * @property {string} [ott]
 */

// #2
/**
 * @typedef {object} preInvoiceObj
 * @property {string} [redirectURL]
 * @property {string} [callUrl]
 * @property {integer} [userId]
 * @property {string} [billNumber]
 * @property {string} [description]
 * @property {string} [deadline]
 * @property {productIdListObj[]} productIdList
 * @property {string} guildCode
 * @property {string} [currencyCode]
 * @property {string} [preferredTaxRate]
 * @property {boolean} [verificationNeeded]
 * @property {string} [ott]
 */

// #3
/**
 * @typedef {object} getInvoiceListObj
 * @property {integer} [id]
 * @property {string} [billNumber]
 * @property {string} [uniqueNumber]
 * @property {integer} [trackerId]
 * @property {string} [fromDate]
 * @property {string} [toDate]
 * @property {boolean} [isCanceled]
 * @property {boolean} [isPayed]
 * @property {boolean} [isClosed]
 * @property {boolean} [isWaiting]
 * @property {string} [guildCode]
 * @property {string} [referenceNumber]
 * @property {integer} [userId]
 * @property {integer} [preferredissuerId]
 * @property {string} [query]
 * @property {integer} [firstId]
 * @property {integer} [lastId]
 * @property {integer} offset
 * @property {integer} [productIdList]
 * @property {integer} [size]
 */

// #4, #13
/**
* @typedef {object} invoiceIdObj
* @property {integer} invoiceId
*/

// #5
/**
 * @typedef {object} sendInvoicePaymentSMSObj
 * @property {integer} invoiceId
 * @property {string} [wallet]
 * @property {string} [callbackUri]
 * @property {string} [redirectUri]
 */

// #6
/**
 * @typedef {object} getInvoiceListByMetadataObj
 * @property {object} metaQuery
 * @property {integer} [offset]
 * @property {integer} [size]
 */

// #7
/**
 * @typedef {object} getInvoiceListAsFileObj
 * @property {integer} [id]
 * @property {string} [billNumber]
 * @property {string} [uniqueNumber]
 * @property {integer} [trackerId]
 * @property {string} [fromDate]
 * @property {string} [toDate]
 * @property {boolean} [isCanceled]
 * @property {boolean} [isPayed]
 * @property {boolean} [isClosed]
 * @property {boolean} [isWaiting]
 * @property {string} [guildCode]
 * @property {string} [referenceNumber]
 * @property {integer} [userId]
 * @property {string} [query]
 * @property {string} [productIdList]
 * @property {integer} lastNRows
 * @property {string} [callbackUrl]
 */

// #8
/**
 * @typedef {object} verifyInvoiceObj
 * @property {integer} id
 * @property {string} [billNumber]
 */

// #9, #11, #12
/**
 * @typedef {object} idObj
 * @property {integer} id
 */

// #10
/**
 * @typedef {object} reduceInvoiceObj
 * @property {integer} id
 * @property {invoiceItemListObj} invoiceItemListObj
 * @property {string} [preferredTaxRate]
 */

// #14
/**
 * @typedef {object} payInvoiceByInvoiceObj
 * @property {integer} creditorInvoiceId
 * @property {integer} debtorInvoiceId
 */

// #15
/**
 * @typedef {object} payInvoiceInFuture
 * @property {integer} invoiceId
 * @property {string} date
 */

// #16 
/**
 * @typedef {object} getExportList
 * @property {integer} offset
 * @property {integer} size
 * @property {integer} [id]
 * @property {string} [statusCode]
 * @property {string} [serviceUrl]
 */

// #17
/**
 * @typedef {object} requestWalletSettlementObj
 * @property {string} ott
 * @property {string} [wallet]
 * @property {string} [firstname]
 * @property {string} [lastname]
 * @property {string} [sheba]
 * @property {number} amount
 * @property {string} [currencyCode]
 * @property {integer} [uniqueId]
 * @property {number} [description]
 */

// #18
/**
 * @typedef {object} requestGuildSettlementObj
 * @property {string} ott
 * @property {string} [firstname]
 * @property {string} [lastname]
 * @property {string} [sheba]
 * @property {number} amount
 * @property {string} [guildCode]
 * @property {string} [currencyCode]
 * @property {integer} [uniqueId]
 * @property {number} [description]
 */

// #19
/**
 * @typedef {object} requestSettlementByToolObj
 * @property {string} ott
 * @property {string} [firstname]
 * @property {string} [lastname]
 * @property {string} toolCode
 * @property {string} toolId
 * @property {number} amount
 * @property {string} guildCode
 * @property {string} [currencyCode]
 * @property {integer} [uniqueId]
 * @property {number} [description]
 */

// #20
/**
* @typedef {object} listSettlementsObj
* @property {string} [statusCode]
* @property {string} [currencyCode]
* @property {string} [fromAmount]
* @property {string} [toAmount]
* @property {string} [fromDate]
* @property {string} [toDate]
* @property {integer} offset
* @property {integer} [uniqueId]
* @property {integer} [size]
*/

// #21
/**
* @typedef {object} addAutoSettlementObj
* @property {string} [firstName]
* @property {string} [lastName]
* @property {string} [sheba]
* @property {string} guildCode
* @property {string} [currencyCode]
* @property {string} [instant]
*/

// #22
/**
* @typedef {object} removeAutoSettlementObj
* @property {string} guildCode
* @property {string} [currencyCode]
*/

// #23
/**
* @typedef {object} getPayInvoiceByWalletLinkObj
* @property {string} invoiceId
* @property {string} [redirectUri]
* @property {string} [callUri]
*/

// #24
/**
* @typedef {object} getPayInvoiceByUniqueNumberLinkObj
* @property {string} uniqueNumber
* @property {string} [redirectUri]
* @property {string} [callUri]
* @property {string} [gateway]
*/

/**
 * @typedef {object} voucherObj
 * @property {integer} count
 * @property {number} amount
 * @property {string} name
 * @property {string} description
 * @property {string} [hash]
 */

/**
* @typedef {object} percentVoucherObj
* @property {integer} count
* @property {number} amount
* @property {string} name
* @property {string} description
* @property {string} discountPercentage
* @property {string} [hash]
*/

// #28
/**
 * @typedef {object} defineCreditVoucherObj
 * @property {string} guildCode
 * @property {string} expireDate
 * @property {Array<voucherObj>} vouchers
 * @property {integer} [limitedConsumerId]
 * @property {string} [currencyCode]
 * @property {string} [token]
 * @property {integer} [tokenissuer]
 */

// #29 
/**
 * @typedef {object} defineDiscountAmountVoucherObj
 * @property {string} guildCode
 * @property {string} expireDate
 * @property {Array<voucherObj>} vouchers
 * @property {Array<integer>} [productId]
 * @property {Array<integer>} [dealerBusinessId]
 * @property {integer} [limitedConsumerId]
 * @property {string} [currencyCode]
 * @property {string} [token]
 * @property {integer} [tokenissuer]
 */

// #30
/**
* @typedef {object} defineDiscountPercentageVoucherObj
 * @property {string} guildCode
 * @property {string} expireDate
 * @property {Array<percentVoucherObj>} vouchers
 * @property {integer} type
 * @property {Array<integer>} [productId]
 * @property {Array<integer>} [dealerBusinessId]
 * @property {integer} [limitedConsumerId]
 * @property {string} [currencyCode]
 * @property {Array<string>} [hash]
 * @property {string} [token]
 * @property {integer} [tokenissuer]
*/

// #31
/**
 * @typedef {object} applyVoucherObj
 * @property {string} ott
 * @property {integer} invoiceId
 * @property {Array<string>} voucherHash
 * @property {string} [token]
 * @property {integer} [tokenissuer]
 */

// #32
/**
 * @typedef {object} getVoucherListObj
 * @property {integer} offset
 * @property {integer} size
 * @property {integer} [consumerId]
 * @property {string} [hash]
 * @property {string} [name]
 * @property {integer} [type]
 * @property {Array<string>} [guildCode]
 * @property {string} [currencyCode]
 * @property {number} [amountFrom]
 * @property {number} [amountTo]
 * @property {number} [discountPercentageFrom]
 * @property {number} [discountPercentageTo]
 * @property {string} [expireDateFrom]
 * @property {string} [expireDateTo]
 * @property {string} [consumDateFrom]
 * @property {string} [consumDateTo]
 * @property {number} [usedAmountFrom]
 * @property {number} [usedAmountTo]
 * @property {boolean} [active]
 * @property {boolean} [used]
 * @property {string} [token]
 * @property {integer} [tokenissuer]
 */

// #33, #34
/**
 * @typedef {object} activationVoucherObj
 * @property {integer} id
 * @property {string} [token]
 * @property {integer} [tokenissuer]
 */


lib/products.js
module.exports = {
  // #1
  issueInvoice: 29796,

  // #2
  createPreInvoice: 30359,

  // #3
  getInvoiceList: 29813,

  // #4
  payInvoice: 29799,

  // #5
  sendInvoicePaymentSMS: 29807,

  // #6
  getInvoiceListByMetadata: 29814,

  // #7
  getInvoiceListAsFile: 29817,

  // #8
  verifyInvoice: 29809,

  // #9
  cancelInvoice: 29810,

  // #10
  reduceInvoice: 29803,

  // #11
  verifyAndCloseInvoice: 29808,

  // #12
  closeInvoice: 29812,

  // #13
  getInvoicePaymentLink: 29815,

  // #14
  payInvoiceByInvoice: 29800,

  // #15
  payInvoiceInFuture: 29802,

  // #16
  getExportList: 29818,

  // #17
  requestWalletSettlement: 29986,

  // #18
  requestGuildSettlement: 29987,

  // #19
  requestSettlementByTool: 29988,

  // #20
  listSettlements: 29989,

  // #21
  addAutoSettlement: 29990,

  // #22
  removeAutoSettlement: 29991,

  // #23 - getPayInvoiceByWalletLink

  // #24 - getPayInvoiceByUniqueNumberLink

  // #25
  issueMultiInvoice: 29798,

  // #26
  reduceMultiInvoice: 29805,

  // #27
  reduceMultiInvoiceAndCashOut: 29806,

  // #28
  defineCreditVoucher: 29975,

  // #29
  defineDiscountAmountVoucher: 29976,

  // #30
  defineDiscountPercentageVoucher: 29977,

  // #31
  applyVoucher: 29811,

  // #32
  getVoucherList: 29979,

  // #33
  deactivateVoucher: 29980,

  // #34
  activateVoucher: 29981,

  // #35
  defineDirectWithdraw: 29982,

  // #36
  directWithdrawList: 29983,

  // #37
  updateDirectWithdraw: 29984,

  // #38
  revokeDirectWithdraw: 29985,

  // #39
  payInvoiceByCredit: 30382,

  // #40
  payAnyInvoiceByCredit: 29821
};


lib/urls.js
module.exports = {
  // #1
  issueInvoice: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #2
  createPreInvoice: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #3
  getInvoiceList: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #4
  payInvoice: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #5
  sendInvoicePaymentSMS: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #6
  getInvoiceListByMetadata: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #7
  getInvoiceListAsFile: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #8
  verifyInvoice: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #9
  cancelInvoice: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #10
  reduceInvoice: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #11
  verifyAndCloseInvoice: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #12
  closeInvoice: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #13
  getInvoicePaymentLink: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #14
  payInvoiceByInvoice: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #15
  payInvoiceInFuture: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #16
  getExportList: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #17
  requestWalletSettlement: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #18
  requestGuildSettlement: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #19
  requestSettlementByTool: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #20
  listSettlements: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #21
  addAutoSettlement: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #22
  removeAutoSettlement: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #23
  getPayInvoiceByWalletLink: {
    baseUrl: 'privateCallAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #24
  getPayInvoiceByUniqueNumberLink: {
    baseUrl: 'privateCallAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #25
  issueMultiInvoice: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #26
  reduceMultiInvoice: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #27
  reduceMultiInvoiceAndCashOut: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // --------------------------------------------- Voucher ---------------------------------------------
  // #28
  defineCreditVoucher: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #29
  defineDiscountAmountVoucher: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #30
  defineDiscountPercentageVoucher: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #31
  applyVoucher: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #32
  getVoucherList: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #33
  deactivateVoucher: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #34
  activateVoucher: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // --------------------------------------------- Direct Debate ---------------------------------------------
  // #35
  defineDirectWithdraw: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #36
  directWithdrawList: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #37
  updateDirectWithdraw: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #38
  revokeDirectWithdraw: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // --------------------------------------------- RECENTLY ADDED ---------------------------------------------
  // #39
  payInvoiceByCredit: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  },

  // #40
  payAnyInvoiceByCredit: {
    baseUrl: 'platformAddress',
    subUrl: '/nzh/doServiceCall',
    method: 'POST'
  }
};
